File: App/src/TestMesoCrystal.cpp
#include "TestMesoCrystal.h"
#include "IsGISAXSTools.h"
#include "Types.h"
#include "Units.h"
#include "GISASExperiment.h"
#include "MultiLayer.h"
#include "MaterialManager.h"
#include "LayerDecorator.h"
#include "ParticleDecoration.h"
#include "MesoCrystal.h"
#include "NanoParticleCrystal.h"
#include "FormFactors.h"
#include "LatticeBasis.h"
#include "MathFunctions.h"
#include "Utils.h"


namespace
{
double testResolutionFunction(double u, double v)
{
    double sigma_u = 0.0002;
    double sigma_v = 0.0002;
    return MathFunctions::IntegratedGaussian(u, 0.0, sigma_u)
            * MathFunctions::IntegratedGaussian(v, 0.0, sigma_v);
}
}

TestMesoCrystal::TestMesoCrystal()
: mp_intensity_output(0)
, mp_sample(0)
{
}

TestMesoCrystal::~TestMesoCrystal()
{
    delete mp_intensity_output;
    delete mp_sample;
}

void TestMesoCrystal::execute()
{
    if (mp_intensity_output) delete mp_intensity_output;
    initializeSample();
    GISASExperiment experiment;
    experiment.setSample(mp_sample);
    experiment.setDetectorParameters(100, 0.3*Units::degree, 0.073
           , 100 , -0.4*Units::degree, 0.066);
//    experiment.setDetectorParameters(5, 2*Units::degree, 3*Units::degree
//           , 5 , 1*Units::degree, 2*Units::degree);
//    experiment.setDetectorResolutionFunction(&testResolutionFunction);
    experiment.setBeamParameters(1.77*Units::angstrom, -0.4*Units::degree, 0.0*Units::degree);
    experiment.setBeamIntensity(1e7);
    experiment.runSimulation();
    double count_before_normalize = experiment.getOutputData()->total();
    experiment.normalize();
    mp_intensity_output = experiment.getOutputDataClone();
    double total_count = mp_intensity_output->total();
    std::cout << "Total count in detector: " << total_count << std::endl;
    std::cout << "Scattered percentage in detector: " << 100*total_count/experiment.getBeam().getIntensity() << std::endl;
    std::cout << "Total count in detector before normalize: " << count_before_normalize << std::endl;
    IsGISAXSTools::drawLogOutputData(*mp_intensity_output, "c1_test_meso_crystal", "mesocrystal",
            "CONT4 Z");
    IsGISAXSTools::writeOutputDataToFile(*mp_intensity_output, Utils::FileSystem::GetHomePath()+"./Examples/MesoCrystals/mesocrystal.ima");
}

void TestMesoCrystal::initializeSample()
{
    delete mp_sample;
    // create mesocrystal
    double meso_width = 300*Units::nanometer;
    double surface_filling_ratio = 0.25;
    double surface_density = surface_filling_ratio/meso_width/meso_width;
//    complex_t n_particle(1.0-1.55e-5, 1.37e-6); // data from Artur
    complex_t n_particle(1.0-2.84e-5, 4.7e-7); // data from http://henke.lbl.gov/optical_constants/getdb2.html
    complex_t avg_n_squared_meso = 0.7886*n_particle*n_particle + 0.2114;
    complex_t n_avg = std::sqrt(surface_filling_ratio*avg_n_squared_meso + 1.0 - surface_filling_ratio);
    complex_t n_particle_adapted = std::sqrt(n_avg*n_avg + n_particle*n_particle - 1.0);
    FormFactorCylinder ff_meso(0.2*Units::micrometer, meso_width);

    // Create multilayer
    MultiLayer *p_multi_layer = new MultiLayer();
    complex_t n_air(1.0, 0.0);
    complex_t n_substrate(1.0-7.57e-6, 1.73e-7);

    const IMaterial *p_air_material = MaterialManager::instance().addHomogeneousMaterial("Air", n_air);
    const IMaterial *p_average_layer_material = MaterialManager::instance().addHomogeneousMaterial("Averagelayer", n_avg);
    const IMaterial *p_substrate_material = MaterialManager::instance().addHomogeneousMaterial("Substrate", n_substrate);
    Layer air_layer;
    air_layer.setMaterial(p_air_material);
    Layer avg_layer;
    avg_layer.setMaterial(p_average_layer_material);
    avg_layer.setThickness(0.2*Units::micrometer);
    Layer substrate_layer;
    substrate_layer.setMaterial(p_substrate_material);
    IInterferenceFunction *p_interference_funtion = new InterferenceFunctionNone();
    ParticleDecoration particle_decoration;
    size_t n_phi_rotation_steps = 241;
    size_t n_alpha_rotation_steps = 7;
    double phi_step = 2.0*M_PI/3.0/n_phi_rotation_steps;
    double phi_start = 0.0;
    double alpha_step = 5.0*Units::degree/n_alpha_rotation_steps;
    double alpha_start = - (n_alpha_rotation_steps/2.0)*alpha_step;
    for (size_t i=0; i<n_phi_rotation_steps; ++i) {
        for (size_t j=0; j<n_alpha_rotation_steps; ++j) {
            Geometry::RotateZ3D transform1(phi_start + i*phi_step);
            Geometry::RotateY3D transform2(alpha_start + j*alpha_step);
            Geometry::Transform3D *p_total_transform = new Geometry::Transform3D(transform1*transform2);
            particle_decoration.addParticle(createMesoCrystal(6.1*Units::nanometer,
                    n_particle_adapted, &ff_meso), p_total_transform, 0.2*Units::micrometer);
        }
    }
    particle_decoration.setTotalParticleSurfaceDensity(surface_density);
    particle_decoration.addInterferenceFunction(p_interference_funtion);
    LayerDecorator avg_layer_decorator(avg_layer, particle_decoration);

    p_multi_layer->addLayer(air_layer);
    p_multi_layer->addLayer(avg_layer_decorator);
    p_multi_layer->addLayer(substrate_layer);

//    LayerDecorator air_decorator(air_layer, particle_decoration);
//    p_multi_layer->addLayer(air_decorator);
//    p_multi_layer->addLayer(substrate_layer);


    mp_sample = p_multi_layer;

    std::cout << "Average layer index: " << n_avg << std::endl;
    std::cout << "Adapted particle index: " << n_particle_adapted << std::endl;
    std::cout << "Substrate layer index: " << n_substrate << std::endl;
}

MesoCrystal* createMesoCrystal(double nanoparticle_radius, complex_t n_particle,
        const IFormFactor* p_meso_form_factor)
{
    Lattice lat = Lattice::createTrigonalLattice(nanoparticle_radius*2.0, nanoparticle_radius*2.0*2.3);
    kvector_t bas_a = lat.getBasisVectorA();
    kvector_t bas_b = lat.getBasisVectorB();
    kvector_t bas_c = lat.getBasisVectorC();
    Particle particle(n_particle, new FormFactorFullSphere(nanoparticle_radius));
    kvector_t position_0 = kvector_t(0.0, 0.0, 0.0);
    kvector_t position_1 = 1.0/3.0*(2.0*bas_a + bas_b + bas_c);
    kvector_t position_2 = 1.0/3.0*(bas_a + 2.0*bas_b + 2.0*bas_c);
    std::vector<kvector_t> pos_vector;
    pos_vector.push_back(position_0);
    pos_vector.push_back(position_1);
    pos_vector.push_back(position_2);
    LatticeBasis basis(particle, pos_vector);
    NanoParticleCrystal npc(basis, lat);
    double relative_sigma_np_radius = 0.3;
    double dw_factor = relative_sigma_np_radius*relative_sigma_np_radius*nanoparticle_radius*nanoparticle_radius/6.0;
    npc.setDWFactor(dw_factor);
    return new MesoCrystal(npc.clone(), p_meso_form_factor->clone());
}

File: Examples/UsageWithPython/ex04_MesoCrystal/test_mesocrystal.py
#
# building meso crystal and running the simulation
#
import sys, os
from numpy import linspace, array, zeros, pi, sqrt
import pylab
from matplotlib.colors import LogNorm
import matplotlib.pyplot as plt

sys.path.append(os.path.abspath(
                os.path.join(os.path.split(__file__)[0],
                '..', '..', '..', 'lib')))

from libScattCore import *

M_PI = pi
m_meso_radius = 1000*nanometer
m_surface_filling_ratio = 0.25
m_meso_height = 0.5*micrometer
m_sigma_meso_height = 4*nanometer
m_sigma_meso_radius = 50*nanometer
m_lattice_length_a = 6.2*nanometer
m_nanoparticle_radius = 4.3*nanometer
m_sigma_nanoparticle_radius = 0.14*nanometer
m_sigma_lattice_length_a = 1.5*nanometer
m_roughness = 1.0*nanometer


# --------------------------------------
# building mesocrystal
# ---------------------------------------
def createMesoCrystal(stacking_radius, n_particle, p_meso_form_factor):
  p_lat = createLattice(stacking_radius)
  bas_a = p_lat.getBasisVectorA()
  bas_b = p_lat.getBasisVectorB()
  bas_c = p_lat.getBasisVectorC()
  ff_sphere = FormFactorSphereGaussianRadius(m_nanoparticle_radius, m_sigma_nanoparticle_radius)
  particle = Particle(n_particle, ff_sphere )
  position_0 = kvector_t(0.0, 0.0, 0.0)
  position_1 = kvector_t(0.0, 0.0, 0.0)
  position_2 = kvector_t(0.0, 0.0, 0.0)

  position_1 = 1.0/3.0*(2.0*bas_a + bas_b + bas_c)
  position_2 = 1.0/3.0*(bas_a + 2.0*bas_b + 2.0*bas_c)
  positions = [ position_0, position_1, position_2 ]
  basis = LatticeBasis()
  basis.addParticle(particle, positions)

  npc = Crystal(basis, p_lat)
  dw_factor = m_sigma_lattice_length_a*m_sigma_lattice_length_a/6.0
  npc.setDWFactor(dw_factor)
  meso = MesoCrystal(npc, p_meso_form_factor)
  return meso


# --------------------------------------
# building lattice
# ---------------------------------------
def createLattice(stacking_radius):
  lattice = Lattice.createTrigonalLattice(stacking_radius*2.0, stacking_radius*2.0*2.3)
  p_result = Lattice(lattice)
  selection_rule = SimpleSelectionRule(-1, 1, 1, 3)
  p_result.setSelectionRule( selection_rule )
  return p_result


# ------------------------
# building sample
# ------------------------
def BuildWholeSample():

  surface_density = m_surface_filling_ratio/M_PI/m_meso_radius/m_meso_radius
  n_particle = complex(1.0-2.84e-5, 4.7e-7)
  avg_n_squared_meso = complex(0.7886*n_particle*n_particle + 0.2114)
  n_avg = complex( sqrt(m_surface_filling_ratio*avg_n_squared_meso + 1.0 - m_surface_filling_ratio) )
  n_particle_adapted = complex( sqrt(n_avg*n_avg + n_particle*n_particle - 1.0) )

  ff_cyl = FormFactorCylinder(m_meso_height, m_meso_radius)
  ff_meso = FormFactorDecoratorDebyeWaller(ff_cyl, m_sigma_meso_height*m_sigma_meso_height/2.0, m_sigma_meso_radius*m_sigma_meso_radius/2.0)

  # Create multilayer
  p_multi_layer = MultiLayer()
  n_air = complex(1.0, 0.0)
  n_substrate = complex(1.0-7.57e-6, 1.73e-7)

  p_air_material = MaterialManager.instance().addHomogeneousMaterial("Air", n_air)
  p_average_layer_material = MaterialManager.instance().addHomogeneousMaterial("Averagelayer", n_avg)
  p_substrate_material = MaterialManager.instance().addHomogeneousMaterial("Substrate", n_substrate)

  air_layer = Layer()
  air_layer.setMaterial(p_air_material);

  avg_layer = Layer()
  avg_layer.setMaterial(p_average_layer_material);
  avg_layer.setThickness(m_meso_height);

  substrate_layer = Layer()
  substrate_layer.setMaterial(p_substrate_material)
  p_interference_funtion = InterferenceFunctionNone()
  particle_decoration = ParticleDecoration()

  n_max_phi_rotation_steps = 180
  n_alpha_rotation_steps = 1

  alpha_step = 5.0*degree/n_alpha_rotation_steps
  alpha_start = - (n_alpha_rotation_steps/2.0)*alpha_step

  phi_step = 2.0*M_PI/3.0/n_max_phi_rotation_steps
  phi_start = 0.0
  for i in range(0, n_max_phi_rotation_steps):
    for j in range(0, n_alpha_rotation_steps):
      transform1 = RotateZ3D(phi_start + i*phi_step)
      transform2 = RotateY3D(alpha_start + j*alpha_step)
      p_total_transform = Transform3D(transform1)
      mesocrystal = createMesoCrystal(m_lattice_length_a, n_particle_adapted, ff_meso)
      particle_decoration.addParticle(mesocrystal, p_total_transform, m_meso_height)

  particle_decoration.setTotalParticleSurfaceDensity(surface_density)
  particle_decoration.addInterferenceFunction(p_interference_funtion)
  avg_layer_decorator = LayerDecorator(avg_layer, particle_decoration)

  roughness = LayerRoughness(m_roughness, 0.3, 500.0*nanometer)

  p_multi_layer.addLayer(air_layer)
  p_multi_layer.addLayer(avg_layer_decorator)
  p_multi_layer.addLayerWithTopRoughness(substrate_layer, roughness)

  print "Average layer index: ", n_avg
  print "Adapted particle index: ", n_particle_adapted
  print "Substrate layer index: ", n_substrate
  return p_multi_layer

# ---------------------------------------------------------------
# building sample, experiment, run calculations and plot output
# ---------------------------------------------------------------
def main():
  # building sample
  sample = BuildWholeSample()
  # running experiment
  ai=0.4
  experiment = GISASExperiment()
  experiment.setSample(sample)
  experiment.setDetectorParameters(256, 0.3*degree, 0.073, 256, -0.4*degree, 0.066) # phi, alpha
  experiment.setBeamParameters(1.77*angstrom, -ai*degree, 0.0*degree)
  experiment.setBeamIntensity(8e12)
  print "wait half of the minute..."
  experiment.runSimulation()
  experiment.normalize()

  # getting output data and making plots
  phi = GetOutputDataAxis(experiment, 0)
  alpha = GetOutputDataAxis(experiment, 1)
  output = GetOutputData(experiment) # [phi][alpha]

  phi = phi*180./3.1415926
  alpha = alpha*180./3.1415926
  output2 = output.swapaxes(0, 1)# to have [alpha][phi] ?
  #im=plt.imshow(output2, norm=LogNorm(100, 1e7))
  #plt.colorbar(im)
  #print( "showing" )
  #plt.show()

  fg = pylab.figure()
  xphi,yalpha = pylab.meshgrid(phi,alpha)
  pylab.pcolormesh(xphi,yalpha,output2,  norm=LogNorm(100, 1e+07) )
  pylab.ylim([alpha.min(),alpha.max()])
  pylab.xlim([phi.min(),phi.max()])
  pylab.colorbar()
  pylab.show()


  
#-------------------------------------------------------------
# main()
#-------------------------------------------------------------
if __name__ == '__main__':
  main()


